src/Collections/Arrays/Arrays.dfy

```diff
@@ -0,0 +1,374 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../Sets/Sets.dfy"
+include "../../Comparison.dfy"
+
+module Arrays{
+
+import opened Sets
+import opened Comparison
+import opened Wrappers
+
+  trait Sorter<T> {
+    ghost const C := Comparison(Cmp)
+
+    function method Cmp(t0: T, t1: T): Cmp
+
+    twostate predicate Identical(arr: array<T>, lo: int, hi: int)
+      requires 0 <= lo <= hi <= arr.Length
+      reads arr
+    {
+      arr[lo..hi] == old(arr[lo..hi])
+    }
+
+    twostate lemma IdenticalSplit(arr: array<T>, lo: int, mid: int, hi: int)
+      requires 0 <= lo <= mid <= hi <= arr.Length
+      requires Identical(arr, lo, hi)
+      ensures Identical(arr, lo, mid)
+      ensures Identical(arr, mid, hi)
+    {
+      assert arr[lo..mid] == arr[lo..hi][..mid-lo];
+    }
+
+    twostate lemma IdenticalShuffled(arr: array<T>, lo: int, hi: int)
+      requires 0 <= lo <= hi <= arr.Length
+      requires Identical(arr, lo, hi)
+      ensures Shuffled(arr, lo, hi)
+    {}
+
+    twostate predicate SameElements(arr: array<T>, lo: int, hi: int)
+      reads arr
+      requires 0 <= lo <= hi <= arr.Length
+    {
+      && Identical(arr, 0, lo)
+      && Shuffled(arr, lo, hi)
+      && Identical(arr, hi, arr.Length)
+    }
+
+    twostate lemma SameElementsExtend(arr: array<T>, lo: int, lo': int, hi': int, hi: int)
+      requires 0 <= lo <= lo' <= hi' <= hi <= arr.Length
+      requires SameElements(arr, lo', hi')
+      ensures SameElements(arr, lo, hi)
+    {
+      assert SameElements(arr, lo', hi');
+      assert Identical(arr, 0, lo') && Shuffled(arr, lo', hi') && Identical(arr, hi', arr.Length);
+      IdenticalSplit(arr, 0, lo, lo'); IdenticalSplit(arr, hi', hi, arr.Length);
+      assert && Identical(arr, 0, lo) && Identical(arr, lo, lo')
+             && Shuffled(arr, lo', hi')
+             && Identical(arr, hi', hi) && Identical(arr, hi, arr.Length);
+      IdenticalShuffled(arr, lo, lo'); IdenticalShuffled(arr, hi', hi);
+      assert && Identical(arr, 0, lo) && Shuffled(arr, lo, lo')
+             && Shuffled(arr, lo', hi')
+             && Shuffled(arr, hi', hi) && Identical(arr, hi, arr.Length);
+      ShuffledConcat(arr, lo, lo', hi'); ShuffledConcat(arr, lo, hi', hi);
+      assert Identical(arr, 0, lo) && Shuffled(arr, lo, hi) && Identical(arr, hi, arr.Length);
+      assert SameElements(arr, lo, hi);
+    }
+
+    twostate predicate Shuffled(arr: array<T>, lo: int, hi: int)
+      requires 0 <= lo <= hi <= arr.Length
+      reads arr
+    {
+      multiset(arr[lo..hi]) == old(multiset(arr[lo..hi]))
```

Comment:
I remember adding this to almost every every function in `MergeSort.dfy` will need to replace that with `shuffled` instead.